### PR TITLE
Fix React 16 compatibility issues; various other patches

### PR DESCRIPTION
**Pull request by bignimbus**
Solves https://github.com/mvasin/react-div-100vh/issues/3 as well as a few other issues.

1. The npm package did not include the `lib` directory (it was in `.gitignore`), which rendered the library unusable as a package.  This branch includes the `lib` directory in source control
2. Iterates the version number in `package` and `package-lock`
3. Falls back to the legacy React ref callback syntax to ensure compatibility with React versions > 15 and < 16.3

### CODE REVIEW

File: src/lib/index.jsx [hunk 1]
```diff
@@ -3,45 +3,72 @@ import React from 'react';
...
+    const node = this.getRef();
+
     Object.entries(this.props.style).forEach(([property, rawValue]) => {
       const match = rvhRegex.exec(rawValue);
-      if (match != null) {
+      if (node && match != null) {
         rvhPropertyFound = true;
         // Guarantee that this only runs for numbers
         const extractedValue = parseFloat(match[0]);
         node.style[property] = extractedValue / 100 * window.innerHeight + 'px';
```
Comment (bignimbus) [inline src/lib/index.jsx]: Why are we manipulating DOM directly here and not using `setState`?

Comment (mvasin) [reply src/lib/index.jsx]: Thanks for the wonderful PR!

I agree, using state should be better. Let's tweak it in a separate commit.

Do you have a reason to use both `createRef()` and callback API for the ref? The callback API isn't going anywhere, and the code will be simpler without `createRef()`.

Also, can you please explain the use case when `lib` folder is required? I used the library is an NPM dependency in a create-react-app project, and I can't say it was unusable, it worked fine.

**Comment by mvasin** [comment]
Thanks for the wonderful PR!

Do you have a reason to use both `createRef()` and callback API for the ref? The callback API isn't going anywhere, and the code will be simpler without `createRef()`.

Also, can you please explain the use case when `lib` folder is required? I used the library is an NPM dependency in a create-react-app project, and I can't say it was unusable, it worked fine.

**Comment by bignimbus** [comment]
@mvasin thanks for pushing on this matter.  I was not able to successfully reproduce the `./lib` issue in a fresh project after downloading the npm version.  Not sure why it came up before.  I should note that I was sourcing from a fork, so maybe that had something to do with it.  How is `/lib` built when this module is published to npm?

Re: ref, I agree.  I can update.

**Comment by mvasin** [comment]
git is primarily for hand-written code, not for generated code like build artefacts, so if you don't have a good reason to include `lib` dir, please remove it.

Before pushing to NPM I do `npm run lib`, that generates `lib` directory.

**Comment by bignimbus** [comment]
Agreed. Thanks.

**Comment by bignimbus** [comment]
@mvasin Updated with the `state` refactor and a few other modifications.  I tried to make detailed commit messages to explain my reasoning.  Thanks for your consideration!

**Comment by mvasin** [comment]
That's a massive effort, thanks!

I've got some corrections here and there, and I hope you're not tired of them. Anyway, there's nothing critical going on here, no one's in a hurry, we just enjoy chatting and learning from each other, right? 😁

I'd love to have it merged and I think we are getting close to it.

**Comment by bignimbus** [comment]
@mvasin sounds good, thanks

**Comment by bignimbus** [comment]
@mvasin this refactor may merit a minor release, maybe `0.3.0` instead of `0.2.1`

File: src/lib/Div100vh.test.js [hunk 2]
```diff
@@ -1,9 +1,58 @@
...
+  describe('#render', () => {
+    it('renders and unmounts without crashing', () => {
+      const div = document.createElement('div');
+      ReactDOM.render(<Div100vh />, div);
+      ReactDOM.unmountComponentAtNode(div);
+    });
+  });
+
+  describe('#computeRvhStyles', () => {
+    test('default behavior', () => {
```
Comment (mvasin) [inline src/lib/Div100vh.test.js]: This test name (as well as the others) is not informative. You have to read the test code in order to figure out how the component should behave. I think this will read better:
```
describe('When no rvh units used in style prop')
   it('has has the same height as window')
```

File: src/lib/Div100vh.test.js [hunk 3]
```diff
@@ -1,9 +1,58 @@
...
-  const div = document.createElement('div');
-  ReactDOM.render(<Div100vh />, div);
-  ReactDOM.unmountComponentAtNode(div);
+const renderComponent = props => renderer.create(
+  <Div100vh {...props} />,
+);
+
+const getDivProps = component => component.root.findByType('div').props;
+
+describe('Div100vh', () => {
```
Comment (mvasin) [inline src/lib/Div100vh.test.js]: The outer `describe('Div100vh')` is redundant; as [CRA documentation says](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#writing-tests):
> You may optionally wrap tests in `describe()` blocks for logical grouping but this is **neither required nor recommended**.

File: src/lib/Div100vh.test.js [hunk 4]
```diff
@@ -1,9 +1,58 @@
...
+      ReactDOM.render(<Div100vh />, div);
+      ReactDOM.unmountComponentAtNode(div);
+    });
+  });
+
+  describe('#computeRvhStyles', () => {
+    test('default behavior', () => {
+      window.innerHeight = 1;
+      const component = renderComponent({});
+      const instance = component.getInstance();
```
[the remark this answers is not in the document]
Comment (bignimbus) [reply src/lib/Div100vh.test.js]: Though `computeRvhStyles` is called on mount, it's called in other contexts as well.  I might be missing something, but wouldn't explicitly invoking the method being tested be preferable to assuming implicit behavior?

File: src/lib/index.jsx [hunk 5]
```diff
@@ -3,45 +3,64 @@ import React from 'react';
...
-        // Guarantee that this only runs for numbers
-        const extractedValue = parseFloat(match[0]);
-        node.style[property] = extractedValue / 100 * window.innerHeight + 'px';
-      }
-    });
-    // Default to height 100vh if no rvh found in style
-    if (!rvhPropertyFound) {
-      node.style.height = window.innerHeight + 'px';
-    }
+    const userDefinedStyle = this.props.style || {
```
[the remark this answers is not in the document]
Comment (bignimbus) [reply src/lib/index.jsx]: Yes, I agree that this method could use another round of iteration for the reasons you specified.  A few thoughts:

* The iteration / parsing over styles could probably be abstracted into a simple helper function (either in a separate module or declared in this one).  It would help with readability and testability
* You are correct that the height setting is redundant, I will fix that

File: src/lib/Div100vh.test.js (hunk 2 above)
Comment (bignimbus) [reply src/lib/Div100vh.test.js]: Thanks for the feedback.  For this case,

> default behavior

might be a bit broad, so I think we're in agreement there.  Perhaps

> when no props are passed to the element

would be a more descriptive name?  Strictly speaking, this test is only for when `props` is empty, not when `props.style` is without `rvh` units.

Would you help me understand what you'd be looking for in the other tests?  I'm having trouble deducing what's

> not informative

about them.  Thank you!

File: src/lib/Div100vh.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mvasin) [reply src/lib/Div100vh.test.js]: You could write a unit test for the function, but then why bother rendering a component? You could go without `react-test-renderer` amenities.

You test it as a component, something that I think is perfectly reasonable; testing a function would be too granular to my mind (but acceptable, if you desire so). And if we test a component, we naturally test how hooks work along with the function, and that's great.

File: src/lib/Div100vh.test.js (hunk 2 above)
Comment (mvasin) [reply src/lib/Div100vh.test.js]: My point is the jest output should read like a documentation. And if a test blows up you have an idea what was expected in the first place in plain English, without reading the test code.

I think this output
```
When style prop in not specified
    ✓ has the same height as window
    ✓ passes the rest of the props through
When style prop does not include rvh units
    ✓ merges given style object with height equal to window
When style prop includes rvh units and other props are set
   ✓ sets height to the percentage of real viewport height and passes through other props
```
is way more informative then
```
    #computeRvhStyles
      ✓ default behavior (11ms)
      ✓ when props are passed, but without style (2ms)
      ✓ when props are passed with style without the rvh unit (2ms)
      ✓ when props are passed with style with the rvh unit (2ms)
```

You're right, I didn't catch the test intention, and that confirms that a test intention must be documented in `describe`/`it` clauses in plain English.

As I was writing potential test descriptions above (please tweak them at your will), I've noticed we don't handle the conflict situation, when a user sets `height` in an object passed to `style` prop. Let's issue a warning.

File: src/lib/index.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mvasin) [reply src/lib/index.jsx]: A separate module is an overkill.

File: src/lib/index.jsx [hunk 6]
```diff
@@ -3,45 +3,64 @@ import React from 'react';
...
   }
 
   componentWillUnmount() {
-    window.removeEventListener('resize', this.computeRvhStyles);
+    window.removeEventListener('resize', this.computeRvhStyles, false);
   }
 
   render() {
-    return <div ref={this.myRef} {...this.props} />;
+    const { style } = this.state;
```
Comment (mvasin) [inline src/lib/index.jsx]: I didn't want to bother you with this opinionated correction, but I personally restructure state when I have to use `this.state` at least twice. In this case it saves no typing and just adds an extra LOC.

File: src/lib/index.jsx [hunk 7]
```diff
@@ -2,46 +2,61 @@ import React from 'react';
 
 const rvhRegex = /(\d+(\.\d*)?)rvh\s*$/;
 
+export const parseStylesAndConvertRvhToPx = (rawValue) => {
```
Comment (bignimbus) [inline src/lib/index.jsx]: I thought this would be a decent solution that addresses your thoughtful comments:

1. It separates the parsing logic into a new function and introduces some variable names that may help the reader grok what's happening
2. It keeps the reducer to preserve immutability but removes bloat from the `updateStyle` method
3. It checks for `NaN` to avoid false positives in the parsing logic

**Comment by mvasin** [comment]
Hi @bignimbus!

I'm on vacation so I can't respond in a timely fashion, sorry about that. But I tried to continue based on what you came up with and will share with you the result of my effort soon.

**Comment by bignimbus** [comment]
@mvasin Looks good!  I saw your comment RE: mocking `getWindowHeight` and pushed an update to address it.
